### PR TITLE
Adding support for subtracting GTIDSets

### PR DESCRIPTION
**Pull request by fulghum**
Needed to support the `gtid_subtract()` built-in function. Also exposes the `ParseMysql56GTIDSet` function so that we can parse GTIDSet strings from GMS. 

Related GMS PR: https://github.com/dolthub/go-mysql-server/pull/2455

### CODE REVIEW

File: go/mysql/mysql56_gtid_set.go [hunk 1]
```diff
@@ -35,6 +35,35 @@ func (iv interval) contains(other interval) bool {
 	return iv.start <= other.start && other.end <= iv.end
 }
 
+// overlaps returns true if any part of |other| overlaps with this interval.
+func (iv interval) overlaps(other interval) bool {
+	return other.start >= iv.start && other.start < iv.end ||
+		other.end <= iv.end && other.end > iv.start ||
+		other.start == iv.start && other.end == iv.end
```
Comment (jycor) [inline go/mysql/mysql56_gtid_set.go:42]:
```suggestion
	return other.start >= iv.start && other.start <= iv.end ||
		other.end <= iv.end && other.end >= iv.start ||
		other.start == iv.start && other.end == iv.end
```

I believe these two tests cases do not match MySQL:
```sql
SELECT GTID_SUBTRACT('3E11FA47-71CA-11E1-9E33-C80AA9429562:21-57', '3E11FA47-71CA-11E1-9E33-C80AA9429562:20-21');
SELECT GTID_SUBTRACT('3E11FA47-71CA-11E1-9E33-C80AA9429562:21-57', '3E11FA47-71CA-11E1-9E33-C80AA9429562:57-58');
```

**Review by jycor — APPROVED** [review]
Missing edge case, otherwise LGTM!

File: go/mysql/mysql56_gtid_set.go (hunk 1 above)
Comment (jycor) [reply go/mysql/mysql56_gtid_set.go:42]: Might not need that last and condition with the boundary changes

File: go/mysql/gtid_set.go [hunk 2]
```diff
@@ -40,6 +40,10 @@ type GTIDSet interface {
 	// Contains returns true if the set is a superset of another set.
 	Contains(GTIDSet) bool
 
+	// Subtract returns a new GTIDSet that is the difference between the receiver
+	// and the supplied GTIDSet.
+	Subtract(GTIDSet) GTIDSet
```
Comment (zachmu) [inline go/mysql/gtid_set.go:45]: Usually called `Difference` for sets

File: go/mysql/mysql56_gtid_set.go [hunk 3]
```diff
@@ -35,6 +35,35 @@ func (iv interval) contains(other interval) bool {
...
+}
+
+// subtract returns a slice of intervals created by subtracting |other| from this interval. If this interval is
+// completely contained by |other|, then nil is returned. If |other| overlaps with the middle of this interval, but
+// not the start or end, then two intervals are returned.
+func (iv interval) subtract(other interval) []interval {
+	if iv.start < other.start {
+		if iv.end > other.end {
+			return []interval{
+				{start: iv.start, end: other.start - 1},
```
Comment (zachmu) [inline go/mysql/mysql56_gtid_set.go:52]: Might mention that this operation is exclusive on both ends

**Review by zachmu — APPROVED** [review]
LGTM!

File: go/mysql/mysql56_gtid_set.go (hunk 1 above)
Comment (fulghum) [reply go/mysql/mysql56_gtid_set.go]: Wow, impressive catch @jycor. Thank you! 🙏  Fixed in 55ce8dbabfb5f9c103eaa7a828d971cd401a73a9 and I'll add those test cases to GMS, too.

File: go/mysql/gtid_set.go (hunk 2 above)
Comment (fulghum) [reply go/mysql/gtid_set.go:47]: I like keeping it as `Subtract` just to match the MySQL built-in function `gtid_subtract()`, but let me know if you feel strongly and I'm happy to change it.

**Pull request merged**